source/val/validation_state.h

```diff
@@ -354,6 +360,10 @@ class ValidationState_t {
   /// Stores the list of decorations for a given <id>
   std::unordered_map<uint32_t, std::vector<Decoration>> id_decorations_;
 
+  /// Stores type declarations which need to be unique (i.e. non-aggregates),
+  /// [opcode, operand words]->result_id.
+  std::map<std::vector<uint32_t>, uint32_t> unique_type_declarations_;
```

Comment:
Why is this a map instead of a set?  You never use the result id value.

Also, why is this an ordered container rather than an unordered container?